### ISSUE TITLE
[Issue] bfm: ioctl failed: `IOCTL_START_VMM`

### ISSUE THREAD

**Issue by Randshot** [Bareflank/hypervisor#461]
Hi,

two of my friends seem to have issues with starting the hypervisor.
It loads (`make load`) just fine, but when they try to start (`make start`) it, the get an error:
```
bfm: ioctl failed: `IOCTL_START_VMM`
```

First I though that it was just a single case, but now another friend of mine encountered the same issue.
For me it works just fine, so I don't get why it won't work for them.
Both of them use Windows 10.

**Comment by rianquinn** [comment]
Check via serial to see what message came up. Likely it's something simple like virtualization extensions are not turned on.

**Comment by Randshot** [comment]
Serial?

**Comment by rianquinn** [comment]
Bareflank supports serial output in addition to `make dump`

**Comment by Randshot** [comment]
Eh, how do I do that?
I tried PortMon, but that didn't let me capture anything.

**Comment by rianquinn** [comment]
You need to identify which port your connected to and then tell Bareflank to use that port:
http://bareflank.github.io/hypervisor/html/index.html#serial

**Comment by nick-kvmhv** [comment]
My situation it must be something related to Windows 10 itself: I have no problem starting it on Ubuntu on the same laptop. Windows 10 gives me identical error when I boot in in KVM or as the main OS. If there is more output in serial than in DebugView, I'll try to dust off my old serial over USB cable and check.

**Comment by rianquinn** [comment]
Yeah, you should get more input as it died safely which means that it knows what the issue is, you just need to get access to the error. Make dump only works when the vCPUs are up and running.

**Comment by nick-kvmhv** [comment]
Before I figure out how to connect that serial cable to a laptop with no serial ports, here is what I get in DbgView when I try to start it:
[BAREFLANK ERROR]: start_vmm failed
[BAREFLANK ERROR]: IOCTL_START_VMM: ioctl_start_vmm failed: 8000000000000030 - ENTRY_ERROR_VMM_START_FAILED

**Comment by rianquinn** [comment]
yeah that just means that it attempted to start but failed. Once option would be to create a global debug ring and only dump from the debug ring, in which case you could get the output from make dump without needing serial

**Comment by nick-kvmhv** [comment]
When running it under KVM, I got this in COM1 output when starting:
----------------------------------------
- Standard Exception Caught            -
----------------------------------------
vector

**Comment by rianquinn** [comment]
Huh... that's weird. that usually means that std::vector generated an exception, likely due to out of bounds. @connojd Have you seen this with KVM? Also... what hardware are you running on?

**Comment by nick-kvmhv** [comment]
I have this: Intel(R) Core(TM) i7-6600U CPU @ 2.60GHz on Lenovo x260 laptop. Keep in mind that It works like a charm under Ubuntu. I have W10 on a SSD USB that I use to boot both directly and under KVM and it gives me this error.

**Comment by rianquinn** [comment]
I will take a look at this on Monday. I have a skylake machine I can test Windows on just to make sure something is broken. @ktemkin do you mind if @connojd takes a quick look at KVM next week?

**Comment by ktemkin** [comment]
That's fine-- @connojd, feel free to spend a bit of time looking into this. :)

**Comment by nick-kvmhv** [comment]
A quick update: it seems to be failing in 
vmcs_intel_x64::write_32bit_guest_state
on the following line:
vmcs::guest_fs_limit::set(state->fs_limit());

I added state->dump(); call before it and here is what I got:
DEBUG: ----------------------------------------
DEBUG: - vmcs_intel_x64_host_vm_state dump    -
DEBUG: ----------------------------------------
DEBUG: 
DEBUG: segment selectors:
DEBUG:     - m_es: 0x2b
DEBUG:     - m_cs: 0x10
DEBUG:     - m_ss: 0x18
DEBUG:     - m_ds: 0x2b
DEBUG:     - m_fs: 0x53
DEBUG:     - m_gs: 0x2b
DEBUG:     - m_ldtr: 0x0
DEBUG:     - m_tr: 0x40
DEBUG: 
DEBUG: segment base:
DEBUG:     - es_base(): 0x0
DEBUG:     - cs_base(): 0x0
DEBUG:     - ss_base(): 0x0
DEBUG:     - ds_base(): 0x0
DEBUG:     - fs_base(): 
----------------------------------------
- Standard Exception Caught            -
----------------------------------------
vector

**Comment by nick-kvmhv** [comment]
methinks I found the problem. Here is the patch that helped:
```
diff --git a/bfvmm/include/intrinsics/gdt_x64.h b/bfvmm/include/intrinsics/gdt_x64.h
index 44a8d82..c59ae3a 100644
--- a/bfvmm/include/intrinsics/gdt_x64.h
+++ b/bfvmm/include/intrinsics/gdt_x64.h
@@ -219,7 +219,7 @@ public:
             m_gdt_reg.base = x64::gdt::base::get();
             m_gdt_reg.limit = x64::gdt::limit::get();
 
-            std::copy_n(m_gdt_reg.base, m_gdt_reg.limit >> 3, std::back_inserter(m_gdt));
+            std::copy_n(m_gdt_reg.base, (m_gdt_reg.limit+1) >> 3, std::back_inserter(m_gdt));
         });
     }
```

**Comment by rianquinn** [comment]
I'm really impressed. Great job debugging this, and I think this is one of the bugs we were seeing with UEFI. Please put together a PR for this so that you get credit for the find/fix. 

Right now we don't have a global debug ring but we really should fix that specifically for issues like this. I will try to draft an RFC that tries to fix this so that make dump works as well as serial.

**Comment by rianquinn** [comment]
@nick-kvmhv This PR should help with debugging a little bit more in v1.2. https://github.com/Bareflank/bfsdk/pull/35. 

Someday I hope to extend the unwinder to provide function names by parsing the debug ELF section. This way, as it's unwinding, it can provide a full stack dump. The unwinder has all of the information it needs, it just has to be given the debug section, and some code to parse the function table. 

For now, the above PR should at least help identify what type of exception was thrown.

**Comment by rianquinn** [comment]
@nick-kvmhv This PR has a bunch of cleanup in it, but one thing I added was a global debug ring, so in the future you should be able to do a `make dump` and get the exception information without needing serial. 

https://github.com/Bareflank/bfvmm/pull/36

**Comment by nick-kvmhv** [comment]
make dump is very nice to have working. Is there any way to use a USB .3.0 debug cable with it? I assume it may require coding some direct USB communication classes? Unfortunately, once in virtual mode, you rarely get "nice" crashes.

**Comment by rianquinn** [comment]
I'd love to add USB 3.0 debug support. Years ago I wrote some of the code needed for USB 2.0 debugging with the special cable that you needed, but my understanding is that USB 3.0 did away with a lot of that. 

And yeah, getting info from VMX can be a pain depending on the issue. I don't have time to write the USB logic right now as I'm working on vAPIC support, but I'd gladly take a PR. Also... let me know if you run into any issues updating your current PR.

**Comment by Randshot** [comment]
Since the issue seems to be fixed with the latest commit (9de33c3c024b48c0259478e7c02b9841fa1e65a5), I will now close this issue.
